pkg/controller/kubelet-config/kubelet_config_bootstrap.go

```diff
@@ -0,0 +1,165 @@
+			}
+
+			kubeletIgnition, logLevelIgnition, autoSizingReservedIgnition, err := generateKubeletIgnFiles(kubeletConfig, originalKubeConfig, userDefinedSystemReserved)
+			if err != nil {
+				return nil, err
+			}
+
+			tempIgnConfig := ctrlcommon.NewIgnConfig()
+			if autoSizingReservedIgnition != nil {
+				tempIgnConfig.Storage.Files = append(tempIgnConfig.Storage.Files, *autoSizingReservedIgnition)
+			}
+			if logLevelIgnition != nil {
+				tempIgnConfig.Storage.Files = append(tempIgnConfig.Storage.Files, *logLevelIgnition)
+			}
+			if kubeletIgnition != nil {
+				tempIgnConfig.Storage.Files = append(tempIgnConfig.Storage.Files, *kubeletIgnition)
+			}
+
+			rawIgn, err := json.Marshal(tempIgnConfig)
+			if err != nil {
+				return nil, err
+			}
+			managedKey, err := generateSuffixManagedKeyKubelet(pool, kubeletConfig, generatedManagedKeys)
+			if err != nil {
+				return nil, err
+			}
+			var (
+				mc *mcfgv1.MachineConfig
+				ok bool
+			)
+			mc, ok = generatedMCs[managedKey]
+			if !ok {
+				mc, err = ctrlcommon.MachineConfigFromIgnConfig(role, managedKey, kubeletIgnition)
+				if err != nil {
+					return nil, err
+				}
+				generatedMCs[managedKey] = mc
+			}
+
+			mc.Spec.Config.Raw = rawIgn
+			mc.SetAnnotations(map[string]string{
+				ctrlcommon.GeneratedByControllerVersionAnnotationKey: version.Hash,
+			})
+			oref := metav1.OwnerReference{
+				APIVersion: controllerKind.GroupVersion().String(),
+				Kind:       controllerKind.Kind,
+			}
+			mc.SetOwnerReferences([]metav1.OwnerReference{oref})
+			res = append(res, mc)
+		}
+	}
+	return res, nil
+}
+
+func generateSuffixManagedKeyKubelet(pool *mcfgv1.MachineConfigPool, cfg *mcfgv1.KubeletConfig, managedKeys map[string]string) (string, error) {
+	var (
+		managedKey string
+		err        error
+	)
+	if len(managedKeys) == 0 {
+		managedKey, err = ctrlcommon.GetManagedKey(pool, nil, "99", "kubelet", getManagedKubeletConfigKeyDeprecated(pool))
+		if err != nil {
+			return "", err
+		}
+		if cfg.Name != "" {
+			managedKeys[cfg.Name] = managedKey
+		} else {
+			managedKeys[managedKey] = managedKey
+		}
+		return managedKey, nil
+	}
+	if cfg.Name != "" {
+		managedKey, ok := managedKeys[cfg.Name]
+		if ok {
+			return managedKey, nil
+		}
+	}
+	// generate new managedKey if the kubelet config name is empty or it is a new kubelet config
+	maxSuffixNum := 0
+	for _, v := range managedKeys {
+		tokens := strings.Split(v, "-")
+		if len(tokens) < 1 {
+			return "", fmt.Errorf("wrong syntax of managedKey: %v", v)
+		}
+		if tokens[len(tokens)-1] == "kubelet" {
+			continue
+		}
+		num, err := strconv.Atoi(tokens[len(tokens)-1])
+		if err != nil {
+			return "", fmt.Errorf("error parsing suffix of managedKey %v", v)
+		}
+		if num > maxSuffixNum {
+			maxSuffixNum = num
+		}
+	}
+	nextSuffixNum := maxSuffixNum + 1
+	if nextSuffixNum > 9 {
+		return "", fmt.Errorf("max number of supported kubelet config (10) has been reached. Please delete old kubelet configs before retrying")
```

Comment:
Q: it's possible to hit the max generated kc in bootstrap?

Could you explain? This doesn't seem desirable?

Reply:
Yes, it's possible to hit the max since it is decided to allow multiple kubeletconfigs. I made the `generateSuffixManagedKeyKubelet` the same as the in-cluster KCC and keep the kubeletconfig and machineconfigpool 1:1 mapping.

Reply:
Since this is during bootstrap, we should probably make the message a little less confusing by having it say e.g. cannot exceed 10 kubeletconfig objects on the system. Please ensure bootstrap kubeletconfig manifests are under 10 per pool and try again.